docs/pallas/tpu/collectives.rst

```diff
@@ -0,0 +1,452 @@
+    pltpu.semaphore_signal(barrier_sem, device_id=neighbor)
+    pltpu.semaphore_wait(barrier_sem)
+
+    dev_id = pltpu.device_id()
+    other_dev_id = 1 - dev_id
+    remote_copy = pltpu.async_remote_copy(
+      src_ref, dst_ref.at[0:8], send_sem, recv_sem, other_dev_id,
+      device_id_type=pltpu.DeviceIdType.LOGICAL,
+    )
+    # remote_copy.wait() for both.
+    remote_copy.wait_send()
+    remote_copy.wait_recv()
+
+    # Both APIs have make_ variants for building a descriptor without starting
+    # it. This allows us to start a DMA in one grid iteration and wait for it
+    # in another.
+    local_copy = pltpu.make_async_copy(src_ref, dst_ref.at[0:8], sem)
+    local_copy.start()
+    local_copy.wait()
+  
+  src_data = pltpu.VMEM((8,128), jnp.float32)
+  dst_data = pltpu.VMEM((16,128), jnp.float32)
+  semaphore = pltpu.SemaphoreType.DMA
+
+  # Inside a kernel:
+  pltpu.run_scoped(
+    move_data, src_data, dst_data, *((semaphore,) * 3)
+  )
+
+  # Or alternatively to use data that is live across iterations:
+  pl.pallas_call(
+      move_data,
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+          num_scalar_prefetch=0,
+          # Could also use other memory spaces and input/outputs.
+          scratch_shapes=[src_data, dst_data, semaphore],
+      ),
+  )()
+
+.. warning::
+   Care must be taken when using DMAs to write to inputs or outputs as the reads and writes to these are pipelined with the execution of the body of your kernel. The pipeline double buffering will cause undefined behavior if you try to carry out a DMA across grid iterations where the block windows/indices of the operands are changing. Modifying inputs is also generally not supported as there is no mechanism for the writes to make it to the source buffer from the current ref of the current block.
+
+Custom Pipelines
+-----------------
+
+Pipeline/Grid Recap
+~~~~~~~~~~~~~~~~~~~~
+
+In :doc:`../quickstart` we saw how when we write a kernel like this:
+
+.. code-block:: python
+
+  def matmul_kernel(x_ref, y_ref, z_ref):
+    z_ref[...] = x_ref[...] @ y_ref[...]
+
+  def matmul(x: jax.Array, y: jax.Array):
+    return pl.pallas_call(
+      matmul_kernel,
+      out_shape=jax.ShapeDtypeStruct((x.shape[0], y.shape[1]), x.dtype),
+      grid=(2, 2),
+      in_specs=[
+        pl.BlockSpec(lambda i, j: (i, 0), (x.shape[0] // 2, x.shape[1])),
+        pl.BlockSpec(lambda i, j: (0, j), (y.shape[0], y.shape[1] // 2))
+      ],
+      out_specs=pl.BlockSpec(
+        lambda i, j: (i, j), (x.shape[0] // 2, y.shape[1] // 2)
+      )
+    )(x, y)
+  k1, k2 = jax.random.split(jax.random.PRNGKey(0))
+  x = jax.random.normal(k1, (1024, 1024))
+  y = jax.random.normal(k2, (1024, 1024))
+  z = matmul(x, y)
+  np.testing.assert_allclose(z, x @ y)
+
+We do a series of pipelined nested loops over the grid indices, slicing and prefetching the inputs for the next iteration and writing the outputs of the previous iteration as we are running the current one.
```

Comment:
Perhaps:
"The `pallas_call` performs ..." instead of "We do". You didn't, Pallas did.